cpp/src/transform/transform.cpp

```diff
@@ -63,6 +65,79 @@ void unary_operation(mutable_column_view output,
              cudf::jit::get_data_ptr(input));
 }
 
+std::vector<std::string> make_template_types(column_view outcol_view, table_view data_view)
+{
+  std::string mskptr_type =
+    cudf::jit::get_type_name(cudf::data_type(cudf::type_to_id<cudf::bitmask_type>())) + "*";
+  std::string offset_type =
+    cudf::jit::get_type_name(cudf::data_type(cudf::type_to_id<cudf::offset_type>()));
+
+  std::vector<std::string> template_types;
+  template_types.reserve(data_view.num_columns() + 1);
+
+  template_types.push_back(cudf::jit::get_type_name(outcol_view.type()));
+  for (auto const& col : data_view) {
+    auto these_types = {cudf::jit::get_type_name(col.type()) + "*", mskptr_type, offset_type};
+    template_types.insert(template_types.end(), these_types);
+  }
+  return template_types;
+}
+
+void generalized_operation(table_view data_view,
+                           std::string const& binary_udf,
+                           data_type output_type,
+                           mutable_column_view outcol_view,
+                           mutable_column_view outmsk_view,
+                           rmm::mr::device_memory_resource* mr)
+{
+  std::vector<std::string> template_types = make_template_types(outcol_view, data_view);
+
+  std::string generic_kernel_name =
+    jitify2::reflection::Template("cudf::transformation::jit::generic_udf_kernel")
+      .instantiate(template_types);
+
+  std::string generic_cuda_source = cudf::jit::parse_single_function_ptx(
+    binary_udf, "GENERIC_OP", cudf::jit::get_type_name(output_type), {0});
+
+  // {size, out_ptr, out_mask_ptr, col0_ptr, col0_mask_ptr, col0_offset, col1_ptr...}
+  std::vector<void*> kernel_args;
+  kernel_args.reserve((data_view.num_columns() * 3) + 3);
+
+  cudf::size_type size   = outcol_view.size();
+  const void* outcol_ptr = cudf::jit::get_data_ptr(outcol_view);
+  const void* outmsk_ptr = cudf::jit::get_data_ptr(outmsk_view);
+  kernel_args.insert(kernel_args.begin(), {&size, &outcol_ptr, &outmsk_ptr});
+
+  std::vector<const void*> data_ptrs;
+  std::vector<cudf::bitmask_type const*> mask_ptrs;
+  std::vector<cudf::offset_type> offsets;
+
+  data_ptrs.reserve(data_view.num_columns());
+  mask_ptrs.reserve(data_view.num_columns());
+  offsets.reserve(data_view.num_columns());
+
+  column_view col;
+  for (int col_idx = 0; col_idx < data_view.num_columns(); col_idx++) {
+    col = data_view.column(col_idx);
+
+    data_ptrs[col_idx] = cudf::jit::get_data_ptr(col);
```

Comment:
This isn't valid anymore. The reserving you're doing isn't making the vector that big, but instead telling it to be ready to be that big. At a minimum, these should be changed to `push_back`, but I still wonder if we can't get this more readable. I've been playing around in [godbolt](https://godbolt.org/z/KrzGncfKe) thinking about this problem. If you don't think this makes it more clear, continue to do it as you have done now. There is nothing wrong with this method outside of the undefined access.

Reply:
I really like the approach in that godbolt link but the problem seems to be when we want to `push_back` onto `pointer_storage_vector`, I'm ending up needing to do something like

```
pointer_storage_vector.push_back(&col.offset())
```
This won't compile because of `lvalue required as unary ‘&’ operand` which makes sense. So the variables have to actually be separately materialized somewhere before I can put them in `pointer_storage_vector`.